modules/telemetry/events/status-events.php

```diff
@@ -0,0 +1,104 @@
+<?php
+
+namespace VIPWorkflow\Modules\Telemetry\Events;
+
+use Automattic\VIP\Telemetry\Tracks;
+use VIPWorkflow\VIP_Workflow;
+use WP_Post;
+
+class Status_Events {
+	/**
+	 * Tracks instance
+	 *
+	 * @var Tracks
+	 */
+	protected Tracks $tracks;
+
+	/**
+	 * Constructor
+	 *
+	 * @param Tracks $tracks The Tracks instance
+	 */
+	public function __construct( Tracks $tracks ) {
+		$this->tracks = $tracks;
+	}
+
+	/**
+	 * Record an event when a post's custom status changes
+	 *
+	 * @param string $new_status The new status
+	 * @param string $old_status The old status
+	 * @param WP_Post $post The post object
+	 */
+	public function record_custom_status_change(
+		string $new_status,
+		string $old_status,
+		WP_Post $post,
+	): void {
+		if ( ! in_array( $post->post_type, VIP_Workflow::instance()->custom_status->get_supported_post_types() ) ) {
+			return;
+		}
+
+		if ( in_array( $new_status, [ $old_status, 'inherit', 'auto-draft', 'publish', 'draft' ] ) ) {
+			return;
+		}
+
+		$this->tracks->record_event( 'post_custom_status_changed', [
+			'new_status' => $new_status,
+			'old_status' => $old_status,
+			'post_id'    => $post->ID,
+		] );
+	}
+
+	/**
+	 * Record an event when a custom status is created
+	 *
+	 * @param string $term The term name
+	 * @param string $slug The term slug
+	 * @param array $args The term arguments
+	 */
+	public function record_add_custom_status(
+		string $term,
+		string $slug,
+		array $args,
```

Comment:
For each of these, can we remove the `$args` parameter, especially if it's not being used?

Reply:
The `do_action` call is passing those arguments though so wouldn't the callback be required to accept it event if it's not used?

Reply:
I'm not sure I understand. In WordPress you're allowed to accept fewer arguments than the callback supplies, so you can register to receive just the first two arguments if desired.

However, I'm just thinking we change the `do_action()` calls to not supply a full `$args` array unless we're recording all of the `$args`, or to only supply targeted data like [suggested below](https://github.com/Automattic/vip-workflow-plugin/pull/21#discussion_r1742441009). The main problems with supplying unnecessary `$args` is:

- If the `do_action()` calling code is modified or refactored, `$args` might not keep the same structure the telemetry part of the code expects. For example when we call the `vw_add_custom_status` action here, we're passing `$args` that can contain `'slug'`, `'position'`, and any other optional arguments given to the function. If that code is refactored in the future, the shape of `$args` might change, which is also coupled with telemetry code, which may also need to be changed. However, if we just pass a `$slug` parameter directly from the `vw_add_custom_status` action call, there's less coupling and room for error.
- Passing unnecessary data to a telemetry library that could potentially contain PII doesn't look very safe, even if we just internally discard most of that data. We should be careful and selective about the exact data we're passing in.

Reply:
> I'm not sure I understand. In WordPress you're allowed to accept fewer arguments than the callback supplies, so you can register to receive just the first two arguments if desired.

Ah thanks for clearing that up. I forgot that that was possible 👍 

> However, I'm just thinking we change the do_action() calls to not supply a full $args array unless we're recording all of the $arg

My thinking when I added the `do_action` calls was that we should also ensure that these hooks are also useful for things other than telemetry(i.e.: expanding the notifications module) thus I wanted to include as much contextual data as possible. That said, I see your point that the `$args` argument could be inconsistent and it would be better to be more specific. I'll make the changes.

Reply:
Thank you! It still looks like we're passing an `$args` along with `vw_add_custom_status` and `vw_delete_custom_status`, can you get those more specific as well?

Reply:
Done in 6a58f5ae591b571ecb8933c58ff152776a7d0f5b